modin/core/dataframe/base/dataframe/dataframe.py

```diff
@@ -0,0 +1,506 @@
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the filter provided.
+        """
+        pass
+
+    @abstractmethod
+    def map(
+        self,
+        function: Callable,
+        axis: Optional[int] = None,
+        dtypes: Optional[str] = None,
+    ) -> "ModinDataframe":
+        """
+        Apply a user-defined function row- wise (or column-wise if axis=1).
+
+        Parameters
+        ----------
+        function : callable
+            The function to map across the dataframe.
+        axis : int, default: None
+            The axis to map over.
+        dtypes : str, default: None
+            The data types for the result. This is an optimization
+            because there are functions that always result in a particular data
+            type, and this allows us to avoid (re)computing it.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe with the map applied.
+
+        Notes
+        -----
+        This does not change the number of rows.
+
+        The user-defined function may increase the number of columns (rows if axis=1),
+            but it should not remove or drop columns and each invocation of the function
+            must generate the same number of new columns.
+        """
+        pass
+
+    @abstractmethod
+    def filter(self, axis: int, condition: Callable) -> "ModinDataframe":
+        """
+        Filter data based on the function provided along the specified axis.
+
+        Parameters
+        ----------
+        axis : int
+            The axis to filter over.
+        condition : callable
+            The function to use for the filter. This function should filter the
+            data itself.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the filter provided.
+        """
+        pass
+
+    @abstractmethod
+    def explode(
+        self,
+        axis: int,
+        function: Callable,
+        result_schema: Optional[Dict[Hashable, type]] = None,
+    ) -> "ModinDataframe":
+        """
+        Explode data based on the function provided along the specified axis.
+
+        Parameters
+        ----------
+        axis : int
+            The axis to expand over.
+        function : callable
+            The function to use to expand the data. This function should accept one
+            row/column, and return multiple.
+        result_schema : dictionary, default: None
+            Mapping from column labels to data types that represents the types of the output dataframe.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the specified axis expanded.
+
+        Notes
+        -----
+        Only one axis can be expanded at a time.
+
+        The user-defined function may increase the number of rows (columns if axis=1),
+                but it should not remove or drop rows.
```

Comment:
```suggestion
        The user-defined function may increase the number of rows (columns if axis=1),
        but it should not remove or drop rows.
```

Possible, it will be a place to indentation warnings during sphinx doc building (generating docs from docstrings stage).